---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  fig.align="center",
  out.width = "100%"
)
```

# SepstratifieR

<!-- badges: start -->
<!-- badges: end -->

The goal of SepstratifieR is to stratify patients with suspected infection into groups with different molecular characteristics. This is done based on the expression level of a small set of genes measured from whole blood.

License: MIT + file LICENSE

## Installation

You can install the development version of SepstratifieR from [GitHub](https://github.com/) with:

``` r
# Install dependencies
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
    
BiocManager::install("batchelor")
BiocManager::install("MatrixGenerics")

# Install SepstratifieR
# install.packages("devtools")
devtools::install_github("jknightlab/SepstratifieR")
```
## Details

### Background
This package is designed to stratify patients with suspected infectious disease into different molecular groups based on a sample of their gene expression from whole blood. These molecular groups are defined based on a small gene signature, and are referred to as sepsis response signature (SRS) groups.

There are three SRS groups, which are as follows:

SRS1 = Composed of sepsis patients with an immunosupressed profile. These individuals are often at high risk of mortality.

SRS2 = Composed of sepsis patients with an immunocompetent profile. These individuals are at lower risk of mortality.

SRS3 = Composed mostly of healthy individuals.


For more information on how SRS groups were originally defined, please refer to the following publications:

https://doi.org/10.1016/S2213-2600(16)00046-1

https://doi.org/10.1164/rccm.201608-1685OC

### The stratification algorithm
To perform stratification on a group of patient samples (i.e. the user's input), SepstratifieR first aligns the input samples to a reference data set containing gene expression profiles from healthy individuals and sepsis patients. This alignment is performed using the mutual nearest neighbours (mNN) algorithm for batch correction. This has the purpose of bringing the predictor variables to the desired scale. 

Next, the samples are classified into SRS groups based on a previously trained random forest model. In addition, each sample is also assigned a quantitative sepsis response score (SRSq) based on a second random forest prediction model. This score (SRSq) goes from 0 to 1. Patients with SRSq close to zero are likely to be healthy, while patients with SRSq close to one are at high risk.

The diagram below describes how the models used by SepstratifieR were built (top panel), as well as how the package's functions perform alignment and classification (bottom panel):

![Schematic diagram of the analysis steps performed by the SepstratifieR package](./man/figures/README-method-diagram.png)

### Input format
The input expected by this function is a data frame object with rows corresponding to individuals/samples and columns corresponding to genes. This data frame must contain at least the following columns: 

1. When using the 7-gene signature defined by Davenport et al:

ENSG00000152219, ENSG00000100814, ENSG00000127334, ENSG00000131355, ENSG00000137337, ENSG00000156414, and ENSG00000115085. 

2. When using the extended 19-gene signature:

ENSG00000144659, ENSG00000103423, ENSG00000135372, ENSG00000079134, ENSG00000135972, ENSG00000087157, ENSG00000165006, ENSG00000111667, ENSG00000182670, ENSG00000097033, ENSG00000165733, ENSG00000103264, ENSG00000152219, ENSG00000100814, ENSG00000127334, ENSG00000131355, ENSG00000137337, ENSG00000156414, and ENSG00000115085. 


If more columns are present, they will simply be ignored.


We recommend that predictor variables have the following units:


Microarray = Background-corrected, VSN-normalized, log-transformed intensity values

RNA-seq = Log-transformed counts per million (i.e. log-cpm)

qPCR = log-transformed relative expression values (i.e. this is equivalent to negative Cq values)


In addition, we recommend removing any technical batch effects from the input data set before using SepstratifieR.

## A brief example
Below is a basic example which shows you how to use this package to stratify a small set of patients into sepsis reponse groups:

```{r example}
# Load package
library(SepstratifieR)

# Load test data set
data(test_data)
head(test_data)

# Stratify patients
predictions <- stratifyPatients(test_data)
```

The results from this prediction look as follows:
```{r example_output}
predictions
```

Futhermore, you can use SepstratifieR's built-in plotting function to check whether the input samples were successfully mapped to the reference set and if there are any clear outliers.

```{r example_plot}
plotAlignedSamples(predictions)
```

## Using a minimal or an extended gene signature for prediction
SepstratifieR enables predictions based on two different gene signatures:

1. A minimal set of 7 genes defined by Davenport et al. in 2016
2. An extended set of 19 genes defined by Cano-Gamez et al. in 2021. This extended set includes the 7 genes proposed by Davenport plus an additional 12 genes derived from integrative analysis of RNA-seq and microarray data for sepsis patients in the GAinS study. 

For further details on the definition of these signatures, please refer to the relevant publications.

The user can specify which gene signature to use for prediction by simply assigning a value to the 'gene_set' parameter. The default behaviour of SepstratifieR is to use the minimal 7-gene signature. This parameter can be modified as illustrated bellow:

```{r stratify_with_minimal_and_extended_gene_signatures, message=FALSE, warning=FALSE}
# Stratify patients based on the 7-gene signature (this is the default option)
predictions <- stratifyPatients(test_data, gene_set = "davenport")

# Stratify patients based on the 19-gene signature
predictions_extended <- stratifyPatients(test_data, gene_set = "extended")
```

IMPORTANT: Note that the extended gene signature was trained using RNA-seq and microarray measurements only. Thus, in its current state this signature should not be used to predict labels from qPCR data.

## Setting the number of mutual nearest neighbours (k)
Perhaps the most important parameter to take into account when performing patient stratification is the number of mutual nearest neighbours in the data alignment step (k). The impact of this parameter on data integration has been previously summarized in the documentation of mNN (https://rdrr.io/github/LTLA/batchelor/man/mnnCorrect.html).

In brief, lower values of 'k' will retain more substructure in the input data, with samples that do not closely resemble the reference set being flagged as outliers (i.e. samples for which no mutual nearest neighbor was found). Conversely, higher values of 'k' will result in a more lenient and aggressive merging, where samples are forced to align with the reference data even if they differ in certain aspects. Higher values of 'k' often result in better performance of data integration, but can also cause outliers to pass undetected, especially when a small group of samples is not well represented in the reference set. 

The authors of mNN suggest that 'k' be set to the expected size for the smallest subpopulation. Based on the proportion of individuals from different sepsis response (SRS) groups previously reported in the context of sepsis patients in intensive care, we recommend that this parameter be set to 20-30% the number of input samples. However, please not that this value might not be ideal if you are using this algorithm in a different patient population.

In the section below we explain how to assess if the predictions obtained with SepstratifieR are robust to the choice of 'k'.

## Running a sensitivity analysis
It is often unclear which value of 'k' is appropriate for a specific analysis. Moreover, if 'k' is low, many samples might be flagged as potential outliers by SepstratifieR, and it can be difficult to distinguish whether they are true outliers or simply samples which remained unmapped in the mNN step. 

The best way to tackle both of these problems is by performing a sensitivity analysis. In this analysis, patient stratification and SRS prediction are repeatedly performed for a range of 'k' values, and the results from each iteration  are compared to each other so as to assess their stability. 

SepstratifieR has a built in function for sensitivity analysis. You can run this function on the same input used for patient stratification, as shown below:
```{r sensitivity_analysis_example}
sensitivity_results <- runSensitivityAnalysis(test_data)
```

The heatmap above clearly shows that the quantitative predictions from SepstratifieR (SRSq) are robust to the choice of 'k', as they do not seem to change. In addition, when computing the variability in predicted SRSq scores across all of the evaluated values of 'k', we obtain a very low variance, as shown in the scatter plot. Finally, because samples flagged as outliers are randomly distributed in the plot and do not show a higher variance compared to the rest of the data, we can be fairly confident that they are not true outliers.

Now let's see how this would look like if a group of outlier samples were present in the input data.

We begin by artificially increasing the expression levels of ARL14EP for the last 30 samples in the data set.
```{r create_outliers}
set.seed(1)
test_data$ENSG00000152219[121:150] <- test_data$ENSG00000152219[121:150] + rnorm(30, mean=8, sd=1)

tail(test_data)
```

These samples should now act as a subgruop of outliers and, indeed, they very clearly separate along PC3.
```{r plot_pca_with_outliers}
preds <- stratifyPatients(test_data, verbose=F)
plotAlignedSamples(preds, pcs=c(1,3), color_by = "mNN_outlier")
```

To confirm this, let's now re-run the sensitivity analysis on this data set. Notice how now the scatter plot clearly shows that the group of samples set as outliers has higher variance in their SRSq estimations. Indeed, most of them are correctly flagged as outliers during mNN alignment. You can also see this in the heatmap, where the SRSq value estimated for the outlier samples abruptly decreases as 'k' increases.
```{r sensitivity_analysis_with_outliers}
sensitivity_results <- runSensitivityAnalysis(test_data)
```

Under this last scenario, the sensitivity analysis suggests we should exclude at least a subset of the samples flagged as outliers. 

## Contact
Eddie Cano-Gamez:  ecg@well.ox.ac.uk
